src/Symfony/Bundle/SecurityBundle/Tests/Functional/SecurityTest.php

```diff
@@ -31,4 +33,142 @@ public function testServiceIsFunctional()
         $this->assertTrue($security->isGranted('ROLE_USER'));
         $this->assertSame($token, $security->getToken());
     }
+
+    public function userWillBeMarkedAsChangedIfRolesHasChangedProvider()
+    {
+        return [
+            [User::class],
+            [UserWithoutEquatable::class],
+        ];
+    }
+
+    /**
+     * @dataProvider userWillBeMarkedAsChangedIfRolesHasChangedProvider
+     */
+    public function testUserWillBeMarkedAsChangedIfRolesHasChanged($userClass)
+    {
+        $client = $this->createClient(['test_case' => 'AbstractTokenCompareRoles', 'root_config' => 'config.yml']);
+        $client->disableReboot();
+
+        /** @var ArrayUserProvider $userProvider */
+        $userProvider = static::$kernel->getContainer()->get('security.user.provider.array');
+        $userProvider->addUser(new $userClass('user1', 'test', ['ROLE_ADMIN']));
+
+        $client->request('POST', '/login', [
+            '_username' => 'user1',
+            '_password' => 'test',
+        ]);
+
+        // user1 has ROLE_ADMIN and can visit secure page
+        $client->request('GET', '/admin');
+        $this->assertEquals(200, $client->getResponse()->getStatusCode());
+
+        // revoking ROLE_ADMIN from user1
+        $userProvider->setUser('user1', new $userClass('user1', 'test', ['ROLE_USER']));
```

Comment:
I would return actual objects from the providers instead.

Reply:
done